src/src-cpp/BjsonDecoder.hpp

```diff
@@ -0,0 +1,165 @@
+ * See http://bjson.org for more.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <string>
+#include <bjson/bjson-common.h>
+#include <bjson/bjson-decode.h>
+
+// -----------------------------------------------------------------------------
+//       Helper macros to wrap pure C calbacks into C++ member calls
+// -----------------------------------------------------------------------------
+
+#define BJSON_CPP_DECODE0(NAME)                              \
+  static int _handlerForPureC_##NAME(void *thiz)             \
+  {                                                          \
+    return (reinterpret_cast<BjsonDecoder*>(thiz))->NAME();  \
+  }
+
+#define BJSON_CPP_DECODE1(NAME, TYPE1)                           \
+  static int _handlerForPureC_##NAME(void *thiz, TYPE1 value)    \
+  {                                                              \
+    return (reinterpret_cast<BjsonDecoder*>(thiz))->NAME(value); \
+  }
+
+#define BJSON_CPP_DECODE2(NAME, TYPE1, TYPE2)                      \
+  static int _handlerForPureC_##NAME(void *thiz, TYPE1 x, TYPE2 y) \
+  {                                                                \
+    return (reinterpret_cast<BjsonDecoder*>(thiz))->NAME(x, y);    \
+  }
+
+
+class BjsonDecoder
+{
+  private:
+
+  // ---------------------------------------------------------------------------
+  // Callback methods called when next token was successfuly decoded.
+  // Overload one or more of below methods to catch decoded values
+  // in stream-like fasion.
+  // ---------------------------------------------------------------------------
+
+  virtual int onNull() {}
+  virtual int onBoolean(int) {}
+  virtual int onInteger(long long) {}
+  virtual int onDouble(double) {}
+  virtual int onNumber(const unsigned char *, size_t) {}
+  virtual int onString(const unsigned char *, size_t) {}
+  virtual int onMapKey(const unsigned char *, size_t) {}
+  virtual int onStartMap() {}
+  virtual int onEndMap() {}
+  virtual int onStartArray() {}
+  virtual int onEndArray() {}
+
+  // ---------------------------------------------------------------------------
+  //                        Internal wrappers (private)
+  // ---------------------------------------------------------------------------
+
```

Comment:
and here starts the "private" part
```suggestion
private:
```